// @insforge/shared-schemas/dist/realtime-api.schema.js
import { z } from 'zod';
import { realtimeChannelSchema, realtimeMessageSchema } from './realtime.schema';
// ============================================================================
// Channel CRUD Schemas
// ============================================================================
// Create Channel
export const createChannelRequestSchema = z.object({
    pattern: z.string().min(1, 'Channel pattern is required'),
    description: z.string().optional(),
    webhookUrls: z.array(z.string().url()).optional(),
    enabled: z.boolean().optional().default(true),
});
export const createChannelResponseSchema = realtimeChannelSchema;
// Update Channel
export const updateChannelRequestSchema = z.object({
    pattern: z.string().min(1).optional(),
    description: z.string().optional(),
    webhookUrls: z.array(z.string().url()).optional(),
    enabled: z.boolean().optional(),
});
export const updateChannelResponseSchema = realtimeChannelSchema;
// Get Channel
export const getChannelResponseSchema = realtimeChannelSchema;
// List Channels
export const listChannelsResponseSchema = z.array(realtimeChannelSchema);
// Delete Channel
export const deleteChannelResponseSchema = z.object({
    message: z.string(),
});
// ============================================================================
// Message Schemas
// ============================================================================
// List Messages
export const listMessagesRequestSchema = z.object({
    channelId: z.string().uuid().optional(),
    eventName: z.string().optional(),
    limit: z.coerce.number().int().min(1).max(1000).optional().default(100),
    offset: z.coerce.number().int().min(0).optional().default(0),
});
export const listMessagesResponseSchema = z.array(realtimeMessageSchema);
// Message Stats
export const messageStatsRequestSchema = z.object({
    channelId: z.string().uuid().optional(),
    since: z.coerce.date().optional(),
});
export const messageStatsResponseSchema = z.object({
    totalMessages: z.number().int().min(0),
    whDeliveryRate: z.number().min(0).max(1),
    topEvents: z.array(z.object({
        eventName: z.string(),
        count: z.number().int().min(0),
    })),
});
// ============================================================================
// Permissions Schemas
// ============================================================================
export const rlsPolicySchema = z.object({
    policyName: z.string(),
    tableName: z.string(),
    command: z.string(),
    roles: z.array(z.string()),
    using: z.string().nullable(),
    withCheck: z.string().nullable(),
});
export const realtimePermissionsResponseSchema = z.object({
    subscribe: z.object({
        policies: z.array(rlsPolicySchema),
    }),
    publish: z.object({
        policies: z.array(rlsPolicySchema),
    }),
});
